fuel/utils.py

```diff
@@ -14,6 +16,347 @@
+    ----------
+    is_list : bool
+        Whether the Subset is a list-based subset (as opposed to a
+        slice-based subset).
+    num_examples : int
+        Number of examples the Subset spans.
+    original_num_examples : int
+        Number of examples in the dataset this subset is part of.
+
+    """
+    def __init__(self, list_or_slice, original_num_examples):
+        self._subset_sanity_check(list_or_slice, original_num_examples)
+        if self._is_list(list_or_slice):
+            list_or_slice = self._beautify_list(list_or_slice)
+        self.list_or_slice = list_or_slice
+        self.original_num_examples = original_num_examples
+
+    def __add__(self, other):
+        if self.original_num_examples != other.original_num_examples:
+            raise ValueError("trying to add two Subset instances with "
+                             "different numbers of original examples, they "
+                             "can't possibly belong to the same dataset")
+        if self.is_list:
+            if other.is_list:
+                return self.__class__(self.list_or_slice + other.list_or_slice,
+                                      self.original_num_examples)
+            else:
+                return self.__class__(self.list_or_slice +
+                                      other[list(range(other.num_examples))],
+                                      self.original_num_examples)
+        else:
+            if other.is_list:
+                return self.__class__(self[list(range(self.num_examples))] +
+                                      other.list_or_slice,
+                                      self.original_num_examples)
+            else:
+                self_sss = self.slice_to_numerical_args(
+                    self.list_or_slice, self.original_num_examples)
+                self_start, self_stop, self_step = self_sss
+                other_sss = self.slice_to_numerical_args(
+                    other.list_or_slice, other.original_num_examples)
+                other_start, other_stop, other_step = other_sss
+                # Single-step slices can be merged into a slice if they
+                # overlap.
+                overlap = not (self_stop < other_start or
+                               self_start > other_stop)
+                if overlap and self_step == other_step == 1:
+                    # In case of overlap, the solution is to choose the
+                    # smallest start value and largest stop value.
+                    return self.__class__(slice(min(self_start, other_start),
+                                                max(self_stop, other_stop),
+                                                self_step),
+                                          self.original_num_examples)
+                # Everything else is transformed into lists before merging.
+                else:
+                    return self.__class__(
+                        self[list(range(self.num_examples))] +
```

Comment:
The idiom `self[list(range(self.num_examples))]` could be factored out as `Subset.as_list_subset`.